Add a README for treetex (#1989)
google/trillian: docs/merkletree/treetex/README.md

Treetex
=======

`treetex` is a tool for drawing/visualising Merkle trees.

> Ensure you have latex installed on your system if you want to this; the tool

Reviewer: ```suggestion
Ensure you have `latex` command installed on your system if you want to use this; the tool
```
Also, rewrap after the modification.

Author: You need more than just the command installed, there's a lot of support files etc. too.

Reviewer: SGTM.

outputs `LaTeX` commands to stdout, and so needs to be used in conjuction with

> that!

Reviewer: ```suggestion
that.
```

For debian-derived systems, `sudo apt install texlive` should do the trick.

Usage
-----

```bash
go run github.com/google/trillian/docs/merkletree/treetex | latex
```
(if you've not used that much tex - that'll create a file called `texput.pdf`)
You can use `okular` (or your favourite PDF viewer) to view it.

> Use `--tree_size` to, er, set the tree size.

Reviewer: ```suggestion
Use `--tree_size` flag to, er, set the tree size.
```

Author: done

Features
--------

> Treetex has a few tricks up its sleeveseses:

Reviewer: Up its Sleevi?

Author: not done :)

> Treetex has a few tricks up its sleeveseses:

Reviewer: ```suggestion
Treetex has a few tricks up its sleeve:
```

Author: done

### Ephemeral nodes

> This tool draws trees with all the leaves at the same level as opposed to floating them up to their real level - mainly because this is how the trees in RFC6962 are drawn.

Reviewer: Wrap?
passim

Author: done

Nodes which aren't yet final are displayed with no border.

### Large trees

> When drawing 'large' tree sizes, set the --megamode_threshold (default is 4) to "collapse" perfect subtrees with at least this many levels.

Reviewer: ```suggestion
When drawing 'large' tree sizes, set the `--megamode_threshold` flag (default is 4) to "collapse" perfect subtrees with at least this many levels.
```

Author: done

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=100023 --megamode_threshold=4
```
![large tree](images/large.png)

### Inclusion proofs

> highlight inclusion proofs

Reviewer: ```suggestion
Highlight an inclusion proof.
```

Author: done

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=23 --inclusion=13
```

![inclusion proof](images/inclusion.png)

### Explicit leaf data
Choose-your-own-~~adventure~~-data.
Takes LaTex commands too.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --leaf_data='entry1,thing too,${\alpha}$,${\frac{bananas_{\beta}}{\delta}}$,{\LaTeX}'
```

![leaf data](images/leafdata.png)

> ### Explain-Merkle-trees-mode

Reviewer: ```suggestion
### Explain-Merkle-trees mode
```

Author: done

Draw diagrams which explain what each node hash is.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
```

![hash mode](images/hashmode.png)

### Compact Range proofs

> Highlight node which make up (multiple) range proofs

Reviewer: ```suggestion
Highlight nodes which make up (multiple) range proofs.
```

```bash

> go run github.com/google/trillian/docs/merkletree/treetex --tree_size=18 --ranges=1:3,7:13

Reviewer: nit (follow-up): I'd replace `7:13` range with `7:14`, for it to be less symmetric and more illustrative. Needs updating the image as well.

Author: done

Treetex
=======

`treetex` is a tool for drawing/visualising Merkle trees.

> Ensure you have LaTeX installed on your system if you want to this; the tool

Reviewer: ```suggestion
Ensure you have `LaTeX` installed on your system if you want to this; the tool
```

Author: done

outputs `LaTeX` commands to stdout, and so needs to be used in conjuction with
that.

For debian-derived systems, `sudo apt install texlive` should do the trick.

Usage
-----

```bash
go run github.com/google/trillian/docs/merkletree/treetex | latex
```
(if you've not used that much tex - that'll create a file called `texput.pdf`)
You can use `okular` (or your favourite PDF viewer) to view it.

Use `--tree_size` flag to, er, set the tree size.

Features
--------
Treetex has a few tricks up its sleeve:

### Ephemeral nodes

This tool draws trees with all the leaves at the same level as opposed to
floating them up to their real level - mainly because this is how the trees in

> RFC6962 are drawn.

Reviewer: ```suggestion
[RFC6962](https://tools.ietf.org/html/rfc6962) are drawn.
```

Author: done

Nodes which aren't yet final are displayed with no border.

### Large trees

> When drawing 'large' tree sizes, set the `--megamode_threshold` flag (default

Reviewer: ```suggestion
When drawing large trees, set the `--megamode_threshold` flag (default
```

is 4) to "collapse" perfect subtrees with at least this many levels.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=100023 --megamode_threshold=4
```
![large tree](images/large.png)

### Inclusion proofs

> Highlight an inclusion proof.

Reviewer: ```suggestion
To highlight an inclusion proof, use the `--inclusion` flag.
```

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=23 --inclusion=13
```

![inclusion proof](images/inclusion.png)

### Explicit leaf data

> Choose-your-own-~~adventure~~-data.

Reviewer: ```suggestion
Choose-your-own-~~adventure~~-data mode. Use the `--leaf_data` flag to specify leaf
data explicitly.
```

Author: done

> Takes LaTex commands too.

Reviewer: nit: Merge this to the previous line.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --leaf_data='entry1,thing too,${\alpha}$,${\frac{bananas_{\beta}}{\delta}}$,{\LaTeX}'
```

![leaf data](images/leafdata.png)

### Explain-Merkle-trees mode

> Draw diagrams which explain what each node hash is.

Reviewer: ```suggestion
To draw diagrams which explain what each node hash is, use the `--node_format` flag.
```

Author: done

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
```

![hash mode](images/hashmode.png)

### Compact Range proofs

> Highlight node which make up (multiple) range proofs

Reviewer: ```suggestion
To highlight nodes which make up (multiple) range proofs, use the `--ranges` flag.
```

Author: done

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=18 --ranges=1:3,7:14
```

> ![compact ranges](images/compactrange.png)

Reviewer: The node 1.6 on this picture should be highlighted.

Author: TODO

Author: Fixed

Treetex
=======

`treetex` is a tool for drawing/visualising Merkle trees.

> Ensure you have `LaTeX` installed on your system if you want to this; the tool

Reviewer: ```suggestion
Ensure you have `LaTeX` installed on your system if you want to use this; the tool
```

Author: done

```
(if you've not used that much tex - that'll create a file called `texput.pdf`)
You can use `okular` (or your favourite PDF viewer) to view it.

Use `--tree_size` flag to, er, set the tree size.

Features
--------
Treetex has a few tricks up its sleeve:

### Ephemeral nodes

This tool draws trees with all the leaves at the same level as opposed to
floating them up to their real level - mainly because this is how the trees in
[RFC6962](https://tools.ietf.org/html/rfc6962) are drawn.
Nodes which aren't yet final are displayed with no border.

### Large trees
When drawing 'large' tree sizes, set the `--megamode_threshold` flag (default
is 4) to "collapse" perfect subtrees with at least this many levels.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=100023 --megamode_threshold=4
```
![large tree](images/large.png)

### Inclusion proofs
Highlight an inclusion proof using the `--inclusion` flag.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=23 --inclusion=13
```

![inclusion proof](images/inclusion.png)

### Explicit leaf data
Choose-your-own-~~adventure~~-data.

> Use the `--leaf_data` flags to specify comma-separated leaf data text explicity.

Reviewer: ```suggestion
Use the `--leaf_data` flag to specify comma-separated leaf data text explicity.
```

Author: done

Takes LaTex commands too.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --leaf_data='entry1,thing too,${\alpha}$,${\frac{bananas_{\beta}}{\delta}}$,{\LaTeX}'
```

![leaf data](images/leafdata.png)

### Explain-Merkle-trees mode
To draw diagrams which explain what each node hash is, use the
`--node_format=hash` flag.

```bash
go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
```

![hash mode](images/hashmode.png)

### Compact Range proofs
To highlight nodes which make up (multiple) range proofs, use the

> `--node_format=address` flag.

Reviewer: ```suggestion
`--ranges` flag.
```

Author: done
